crates/lang_srv/src/registry.rs

```diff
@@ -1,75 +1,191 @@
+        Self {
+            info: latest_doc.doc_info.clone(),
+            latest_document: OnceLock::from(latest_doc),
+            last_good_document,
+        }
+    }
 }
 
 #[derive(Debug, Default)]
 pub(crate) struct Registry {
-    documents: HashMap<Url, AnalyzedDocument>,
+    documents: Mutex<HashMap<Url, DocumentPair>>,
 }
 
 impl Registry {
-    pub fn apply_change(&mut self, change: DocumentChange) {
-        match change {
-            DocumentChange::Modified(url, source) => {
-                let GlobalAnalysis { documents } = GlobalAnalysis::new(url, source);
-
-                // Only replace the set of documents and all dependencies that were re-analyzed.
-                // Note that this is actually the opposite of what we want - in truth we want to
-                // re-evaluate all dependents!
-                for document in documents {
-                    let url = document.url().clone();
-                    self.documents.insert(url.clone(), document);
+    pub async fn get_latest_version(&self, url: &Url) -> Option<i32> {
+        self.documents
+            .lock()
+            .await
+            .get(&url)
+            .map(|x| x.info.version)
+    }
+
+    fn update_document<'a>(
+        documents: &mut MutexGuard<'a, HashMap<Url, DocumentPair>>,
+        document: Arc<AnalyzedDocument>,
+    ) {
+        let url = document.url().clone();
+        match documents.get_mut(&url) {
+            Some(old_doc) => {
+                if document.type_checked() {
+                    *old_doc = DocumentPair::new(document.clone(), document);
+                } else {
+                    debug!(
+                        "Document typechecking failed at version {:?}, not updating last_good_document",
+                        &document.doc_info.version
+                    );
+                    *old_doc = DocumentPair::new(document, old_doc.last_good_document.clone());
                 }
             }
-            DocumentChange::Closed(_url) => {
-                // Do nothing.
+            None => {
+                documents.insert(url.clone(), DocumentPair::new(document.clone(), document));
+            }
+        }
+    }
+
+    pub async fn apply_changes<'a>(&self, analysed_docs: Vec<AnalyzedDocument>, updating_url: Url) {
+        let mut documents = self.documents.lock().await;
+        debug!(
+            "finised doc analysis for doc: {:?}",
+            updating_url.to_string()
+        );
+
+        for document in analysed_docs {
+            let document = Arc::new(document);
+            //Write the newly analysed document into the partial document that any request requiring the latest document will be waiting on
+            if document.doc_info.url == updating_url {
+                documents
+                    .get_mut(&updating_url)
+                    .map(|a| a.latest_document.set(document.clone()).unwrap());
             }
+            Registry::update_document(&mut documents, document);
```

Comment:
We only want to do this if the locked document info version == the analyzed document version, right?

Also, code style thing, I would put the first update and the code in the `update_document` call in the same place, so there aren't disjoint updates.

Reply:
Well we don't call `apply_changes` unless the version is the latest, see `server.rs:169` . I agree, I've moved that code too